cpp/src/arrow/engine/substrait/serde_test.cc

```diff
@@ -2082,43 +2103,121 @@ TEST(Substrait, BasicPlanRoundTrippingEndToEnd) {
+  const std::string file_name = "serde_project_test.arrow";
+
+  ASSERT_OK_AND_ASSIGN(auto tempdir,
+                       arrow::internal::TemporaryDir::Make("substrait-tempdir-project-"));
+  ASSERT_OK_AND_ASSIGN(auto file_path, tempdir->path().Join(file_name));
+  std::string file_path_str = file_path.ToString();
+
+  WriteIpcData(file_path_str, filesystem, table);
+
+  std::vector<fs::FileInfo> files;
+  const std::vector<std::string> f_paths = {file_path_str};
+
+  for (const auto& f_path : f_paths) {
+    ASSERT_OK_AND_ASSIGN(auto f_file, filesystem->GetFileInfo(f_path));
+    files.push_back(std::move(f_file));
   }
-  ASSERT_OK_AND_ASSIGN(auto rnd_trp_table,
-                       GetTableFromPlan(roundtripped_filter, exec_context, dummy_schema));
-  EXPECT_TRUE(expected_table->Equals(*rnd_trp_table));
+
+  ASSERT_OK_AND_ASSIGN(auto ds_factory, dataset::FileSystemDatasetFactory::Make(
+                                            filesystem, std::move(files), format, {}));
+  ASSERT_OK_AND_ASSIGN(auto dataset, ds_factory->Finish(dummy_schema));
+
+  auto scan_options = std::make_shared<dataset::ScanOptions>();
+  scan_options->projection = compute::project({}, {});
+  compute::Expression project_fp_expr = compute::call(
+      "add", {compute::field_ref(FieldPath({1})), compute::field_ref(FieldPath({2}))});
+
+  std::vector<compute::Expression> project_fp_exprs = {project_fp_expr};
+
+  auto declarations = compute::Declaration::Sequence(
+      {compute::Declaration(
+           {"scan", dataset::ScanNodeOptions{dataset, scan_options}, "s"}),
+       compute::Declaration(
+           {"project", compute::ProjectNodeOptions{project_fp_exprs}, "p"})});
+
+  // adding the project expression field to schema
+  auto project_schema = schema({field("add(shared,distinct)", int32())});
+
+  ASSERT_OK_AND_ASSIGN(auto expected_table,
+                       GetTableFromPlan(declarations, exec_context, project_schema));
+
+  std::shared_ptr<ExtensionIdRegistry> sp_ext_id_reg = MakeExtensionIdRegistry();
+  ExtensionIdRegistry* ext_id_reg = sp_ext_id_reg.get();
+  ExtensionSet ext_set(ext_id_reg);
+
+  ASSERT_OK_AND_ASSIGN(auto serialized_plan, SerializePlan(declarations, &ext_set));
+
+  ASSERT_OK_AND_ASSIGN(
+      auto sink_decls,
+      DeserializePlans(
+          *serialized_plan, [] { return kNullConsumer; }, ext_id_reg, &ext_set));
+  // assert project declaration
+  const auto& roundtripped_emit_project =
+      std::get<compute::Declaration>(sink_decls[0].inputs[0]);
+
+  // assert emit projection declaration
+  // emit node used in the serialization outputs the 4th column since the
+  // 0, 1, 2 original columns from the input data and 4th column being the
+  // projected column.
+  auto expec_project_expr = {compute::field_ref(FieldPath({4}))};
+  AssertProjectRelation(roundtripped_emit_project, expec_project_expr, project_schema);
+
+  // assert projection declaration
+  auto subs_project_expr = {compute::field_ref(FieldPath({0})),
```

Comment:
doubtful about this, WDYT?

Reply:
This seems correct for the first project node.  See previous comment.